### PR TITLE
cmd, crypto, keystore: add account subcommand and key generation functionality

### LINKED ISSUE #404: commands to generate, import keys
need to add a command that allows for keys to be generated/imported and saved in the <s>db</s> keystore directory as encrypted keystore files

something like this to generate a new account, which will generate a new keypair, encrypt the private key, save the encrypted private key to a file (name=public address?) 
```
gossamer account --generate --type=sr25519
> type password to encrypt new keystore: ****
```

something like this to import a key (assuming file is encrypted json)
```
gossamer account --import key.json
```

### PR DESCRIPTION
**Pull request by noot**
## Changes
- implemented account subcommand
- account subcommand has `--generate` flag to generate a new key, can do `gossamer account --generate --type=keytype` to specify sr25519 or ed25519 keys 
- has `--list` flag to list all keys in keystore
- has `--import` flag, basically moves file to gossamer keystore directory and checks that it's the correct format

## Tests:
```
go test ./...
```

### Issues:
closes #404

### CODE REVIEW

File: keystore/keystore_test.go [hunk 1]
```diff
@@ -73,12 +84,54 @@ func TestEncryptAndDecryptFromFile(t *testing.T) {
...
+	if !reflect.DeepEqual(priv, res) {
+		t.Fatalf("Fail: got %v expected %v", res, priv)
+	}
+}
+
+func TestEncryptAndDecryptFromFile_Sr25519(t *testing.T) {
+	filename := "./test_key"
+	password := []byte("noot")
+
+	fp, err := filepath.Abs(filename)
```
Comment (ansermino) [inline keystore/keystore_test.go]: You could save a lot of code by using helper functions for things like creating files

File: cmd/gossamer/account.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
+package main
```
Comment (ansermino) [inline cmd/gossamer/account.go]: This file could benefit from more comments

File: cmd/gossamer/account.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	if keygen := ctx.Bool(utils.GenerateFlag.Name); keygen {
+		log.Info("generating keypair...")
+
+		keytype := ""
+		// check if --type flag is set
+		if flagtype := ctx.String(utils.AccountTypeFlag.Name); flagtype != "" {
+			// check if keytype is ed25519 or sr25519
+			if flagtype == "sr25519" || flagtype == "ed25519" {
```
[the remark this answers is not in the document]
Comment (noot) [reply cmd/gossamer/account.go]: good move, I will do that

File: cmd/gossamer/account.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	if keylist := ctx.Bool(utils.ListFlag.Name); keylist {
+		_, err = listKeys(datadir)
+		if err != nil {
+			log.Error("list error", "error", err)
+			os.Exit(1)
+		}
+	}
+}
+
+func importKey(filename, datadir string) (string, error) {
```
[the remark this answers is not in the document]
Comment (noot) [reply cmd/gossamer/account.go]: I want to validate that the contents of the imported file are actually a key, which is why I read it and rewrite it. I guess I could just read it and then copy it over

File: crypto/keypair.go [hunk 5]
```diff
@@ -19,11 +20,12 @@ type PrivateKey interface {
 	Decode([]byte) error
 }
 
-func DecodePrivateKey(in []byte) (PrivateKey, error) {
-	priv, err := NewEd25519PrivateKey(in)
-	if err != nil {
-		return nil, err
+func DecodePrivateKey(in []byte, keytype string) (priv PrivateKey, err error) {
+	if keytype == "ed25519" {
```
[the remark this answers is not in the document]
Comment (noot) [reply crypto/keypair.go]: done

File: cmd/gossamer/account.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+	"github.com/ChainSafe/gossamer/cmd/utils"
+	"github.com/ChainSafe/gossamer/crypto"
+	"github.com/ChainSafe/gossamer/keystore"
+
+	log "github.com/ChainSafe/log15"
+	"github.com/urfave/cli"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+func handleAccounts(ctx *cli.Context) {
```
[the remark this answers is not in the document]
Comment (noot) [reply cmd/gossamer/account.go]: okay I have changed it

File: keystore/keystore_test.go (hunk 1 above)
Comment (noot) [reply keystore/keystore_test.go]: I have added a helper func and cleaned up tests

File: crypto/keypair.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ansermino) [reply crypto/keypair.go]: Please use constants here

File: cmd/gossamer/account.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+	}
+
+	files, err := ioutil.ReadDir(keystorepath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read keystore dir: %s", err)
+	}
+
+	keys := []string{}
+
+	for _, f := range files {
```
Comment (ansermino) [inline cmd/gossamer/account.go]: This prints all files not just keys

File: cmd/gossamer/account.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+		if err != nil {
+			return "", err
+		}
+
+		if _, err := os.Stat(keystorepath); os.IsNotExist(err) {
+			err = os.Mkdir(keystorepath, os.ModePerm)
+			if err != nil {
+				return "", err
+			}
+		}
```
Comment (ansermino) [inline cmd/gossamer/account.go]: Duplicate code below, please simplify

File: cmd/gossamer/account.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+		return keystorepath, nil
+	}
+
+	// datadir not specified, return ~/.gossamer/keystore as absolute path
+	home, err := os.UserHomeDir()
```
Comment (ansermino) [inline cmd/gossamer/account.go]: This can be simplified to: `cfg.DefaultDataDir()`

File: crypto/keypair.go [hunk 10]
```diff
@@ -19,11 +20,12 @@ type PrivateKey interface {
 	Decode([]byte) error
 }
 
-func DecodePrivateKey(in []byte) (PrivateKey, error) {
-	priv, err := NewEd25519PrivateKey(in)
-	if err != nil {
-		return nil, err
+func DecodePrivateKey(in []byte, keytype string) (priv PrivateKey, err error) {
```
Comment (ansermino) [inline crypto/keypair.go]: Why is this necessary if keys have a `Decode` method? Instantiating the type doesn't seem any more cumbersome than specifying the type as a string.

Comment (noot) [reply crypto/keypair.go]: it's used when decoding a private key from an encrypted keystore, since in that case the type is saved as a string inside the file. when we decode the key, it's just raw bytes, we don't have a decode method for it.

Comment (noot) [reply crypto/keypair.go]: see DecryptPrivateKey in keystore/keystore.go

File: cmd/gossamer/account.go (hunk 9 above)
Comment (noot) [reply cmd/gossamer/account.go]: nice, done

File: cmd/gossamer/account.go (hunk 8 above)
Comment (noot) [reply cmd/gossamer/account.go]: done

File: crypto/keypair.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (noot) [reply crypto/keypair.go]: ok done

File: cmd/gossamer/account.go (hunk 7 above)
Comment (noot) [reply cmd/gossamer/account.go]: okay, I have changed it to check for the .key extension

File: crypto/keypair.go (hunk 10 above)
Comment (noot) [reply crypto/keypair.go]: I have made the updates that add KeyType

File: cmd/utils/flags.go [hunk 11]
```diff
@@ -81,3 +91,32 @@ var (
...
+	}
+	ImportFlag = cli.StringFlag{
+		Name:  "import",
+		Usage: "Import encrypted keystore file generated with gossamer",
+	}
+	ListFlag = cli.BoolFlag{
+		Name:  "list",
+		Usage: "List node keys",
+	}
+	Ed25519Flag = cli.StringFlag{
```
Comment (ansermino) [inline cmd/utils/flags.go]: Theses should be bools not strings

Comment (noot) [reply cmd/utils/flags.go]: fixed

**Comment by noot** [comment]
@ryanchristo  I have addded a better description of usage for `--help`. looking into printing the help info for no flags
